File: src/main/java/com/addiction/global/security/config/SecurityConfig.java
package com.addiction.global.security.config;

import java.util.Arrays;
import java.util.Collections;

import org.springframework.context.annotation.Bean;
import org.springframework.context.annotation.Configuration;
import org.springframework.security.config.annotation.web.builders.HttpSecurity;
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity;
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer;
import org.springframework.security.config.annotation.web.configurers.HeadersConfigurer;
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder;
import org.springframework.security.web.SecurityFilterChain;
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter;
import org.springframework.security.web.util.matcher.AntPathRequestMatcher;
import org.springframework.security.web.util.matcher.RequestMatcher;
import org.springframework.web.cors.CorsConfiguration;
import org.springframework.web.cors.CorsConfigurationSource;

import com.addiction.jwt.JwtTokenProvider;
import com.addiction.jwt.filter.JwtAuthenticationFilter;
import com.addiction.jwt.filter.JwtExceptionHandlerFilter;
import com.addiction.user.users.entity.enums.Role;

import lombok.RequiredArgsConstructor;

@RequiredArgsConstructor
@Configuration
@EnableWebSecurity
public class SecurityConfig {

	private final JwtTokenProvider jwtTokenProvider;

	@Bean
	public BCryptPasswordEncoder encoder() {
		return new BCryptPasswordEncoder();
	}

	CorsConfigurationSource corsConfigurationSource() {
		return request -> {
			CorsConfiguration config = new CorsConfiguration();
			config.setAllowedHeaders(Collections.singletonList("*"));
			config.setAllowedMethods(Collections.singletonList("*"));
            config.setAllowedOrigins(Arrays.asList(
                    "https://www.quitmate.co.kr",
                    "https://quitmate.co.kr",
                    "http://localhost:8081"
            ));
			config.setAllowCredentials(true);
			return config;
		};
	}

	@Bean
	protected SecurityFilterChain securityFilterChain(HttpSecurity http) throws Exception {
		http
			.csrf(AbstractHttpConfigurer::disable)
			.cors(corsConfigurer -> corsConfigurer.configurationSource(corsConfigurationSource()))

			.headers((headers) -> headers.frameOptions(HeadersConfigurer.FrameOptionsConfig::disable))

			.authorizeHttpRequests(auth -> auth
				.requestMatchers(getPublicMatchers()).permitAll()
				.anyRequest().hasAnyAuthority(Role.USER.name(), Role.ADMIN.name())
			)

			.logout((logout) -> logout
				.logoutSuccessUrl("/"))

			.addFilterBefore(new JwtAuthenticationFilter(jwtTokenProvider), UsernamePasswordAuthenticationFilter.class)
			.addFilterBefore(new JwtExceptionHandlerFilter(),
				JwtAuthenticationFilter.class); //JwtAuthenticationFilter에서 뱉은 에러를 처리하기 위한 Filter

		return http.build();
	}

	private RequestMatcher[] getPublicMatchers() {
		return new RequestMatcher[] {
			new AntPathRequestMatcher("/api/v1/jwt/**"),
			new AntPathRequestMatcher("/api/v1/auth/**"),
			new AntPathRequestMatcher("/api/v1/user/**"),
			new AntPathRequestMatcher("/docs/**"),
			new AntPathRequestMatcher("/health-check")
		};
	}
}

File: src/main/java/com/addiction/global/security/config/SecurityConfig.kt
package com.addiction.global.security.config

import com.addiction.jwt.JwtTokenProvider
import com.addiction.jwt.filter.JwtAuthenticationFilter
import com.addiction.jwt.filter.JwtExceptionHandlerFilter
import com.addiction.user.users.entity.enums.Role
import org.springframework.context.annotation.Bean
import org.springframework.context.annotation.Configuration
import org.springframework.security.config.annotation.web.builders.HttpSecurity
import org.springframework.security.config.annotation.web.configuration.EnableWebSecurity
import org.springframework.security.config.annotation.web.configurers.AbstractHttpConfigurer
import org.springframework.security.crypto.bcrypt.BCryptPasswordEncoder
import org.springframework.security.web.SecurityFilterChain
import org.springframework.security.web.authentication.UsernamePasswordAuthenticationFilter
import org.springframework.security.web.util.matcher.AntPathRequestMatcher
import org.springframework.security.web.util.matcher.RequestMatcher

@Configuration
@EnableWebSecurity
class SecurityConfig(
    val jwtTokenProvider: JwtTokenProvider
) {

    @Bean
    fun encoder(): BCryptPasswordEncoder {
        return BCryptPasswordEncoder()
    }

    @Bean
    protected fun securityFilterChain(http: HttpSecurity): SecurityFilterChain {
        http
            .csrf(AbstractHttpConfigurer<*, *>::disable)
            .headers { headers ->
                headers.contentSecurityPolicy { policy ->
                    policy.policyDirectives("frame-ancestors 'self'")
                }
            }
            .authorizeHttpRequests { auth ->
                auth
                    .requestMatchers(*getPublicMatchers()).permitAll()
                    .anyRequest().hasAnyAuthority(Role.USER.name, Role.ADMIN.name)
            }
            .logout { logout -> logout.logoutSuccessUrl("/") }
            .addFilterBefore(JwtAuthenticationFilter(jwtTokenProvider), UsernamePasswordAuthenticationFilter::class.java)
            .addFilterBefore(JwtExceptionHandlerFilter(), JwtAuthenticationFilter::class.java) // JwtAuthenticationFilter에서 발생한 에러를 처리하기 위한 필터

        return http.build()
    }

    private fun getPublicMatchers(): Array<RequestMatcher> {
        return arrayOf(
            AntPathRequestMatcher("/api/v1/jwt/**"),
            AntPathRequestMatcher("/api/v1/auth/**"),
            AntPathRequestMatcher("/docs/**"),
            AntPathRequestMatcher("/health-check"),
            AntPathRequestMatcher("/favicon.ico")
        )
    }

}
